### ISSUE TITLE
PM2 deploy setup|update does not work when repository URL is changed

### ISSUE THREAD

**Issue by gajus** [Unitech/pm2#2185]
our repository URL has changed.

I have tried using:

``` bash
$ pm2 deploy ecosystem-production.json production
--> Deploying to production environment
--> on host 149.202.81.6
  ○ deploying origin/master
  ○ executing pre-deploy-local
  ○ hook pre-deploy
  ○ fetching updates
Fetching origin
ERROR: Account `applaudience' is disabled. Please ask the owner to check their account.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Could not fetch origin

  fetch failed

Deploy failed
```

From this I gather PM2 is somehow using the old repository URL.

I have tried using:

``` bash
$ pm2 deploy ecosystem-production.json production setup
--> Deploying to production environment
--> on host 149.202.81.6
  ○ running setup
  ○ cloning git@gitlab.com:applaudience/static-upload-app.git
fatal: destination path '/var/www/static-upload-app/source' already exists and is not an empty directory.

  failed to clone

Deploy failed
```

It now uses correct git URL, but it cannot finish the task because the path already exists. Adding `--force` flag does not make a difference:

``` bash
$ pm2 deploy ecosystem-production.json production setup --force
--> Deploying to production environment
--> on host 149.202.81.6
  ○ running setup
  ○ cloning git@gitlab.com:applaudience/static-upload-app.git
fatal: destination path '/var/www/static-upload-app/source' already exists and is not an empty directory.

  failed to clone

Deploy failed
```

The only workaround I found was to ssh to the server, delete the project and then setup the project again.

**Comment by imink** [comment]
You may first run
pm2 deploy ecosystem-production.json production setup // used to set up all the directories
and then
pm2 deploy ecosystem-production.json production // used to git pull

**Comment by noushad-pp** [comment]
I'm also facing this issue.
```
deploy_dev:
    stage: deploy
    script:
        - echo "====== Deploy to dev server ======"
        - apk update && apk upgrade
        - apk add git openssh bash
        # Add target server`s secret key
        - mkdir ~/.ssh
        - echo $TARGET_SERVER_KEY_TEST_BASE64 | base64 -d > ~/.ssh/id_rsa
        - chmod 700 ~/.ssh && chmod 600 ~/.ssh/*
        - echo "Test ssh connection"
        - ssh -o StrictHostKeyChecking=no -T "$TARGET_SERVER_USER_TEST@$TARGET_SERVER_HOST_TEST"
        # Deploy
        - echo "Setup target server directories"
        - pm2 deploy ecosystem.config.js dev setup 2>&1 || true
        - echo "installing dependencies..."
        - yarn install
        - echo "building..."
        - yarn build --ignore-engines
        - echo "PM2 deploying..."
        # bash:npm not a command issue for nvm installations https://github.com/Unitech/pm2-deploy/issues/41
        - pm2 deploy ecosystem.config.js dev
    environment:
        name: deploying
    only:
        - deploy_dev
```
the error is
```
$ pm2 deploy ecosystem.config.js dev setup 2>&1 || true

                        -------------

__/\\\\\\\\\\\\\____/\\\\____________/\\\\____/\\\\\\\\\_____
 _\/\\\/////////\\\_\/\\\\\\________/\\\\\\__/\\\///////\\\___
  _\/\\\_______\/\\\_\/\\\//\\\____/\\\//\\\_\///______\//\\\__
   _\/\\\\\\\\\\\\\/__\/\\\\///\\\/\\\/_\/\\\___________/\\\/___
    _\/\\\/////////____\/\\\__\///\\\/___\/\\\________/\\\//_____
     _\/\\\_____________\/\\\____\///_____\/\\\_____/\\\//________
      _\/\\\_____________\/\\\_____________\/\\\___/\\\/___________
       _\/\\\_____________\/\\\_____________\/\\\__/\\\\\\\\\\\\\\\_
        _\///______________\///______________\///__\///////////////__


                          Community Edition

            Production Process Manager for Node.js applications
                     with a built-in Load Balancer.


                Start and Daemonize any application:
                $ pm2 start app.js

                Load Balance 4 instances of api.js:
                $ pm2 start api.js -i 4

                Monitor in production:
                $ pm2 monitor

                Make pm2 auto-boot at server restart:
                $ pm2 startup

                To go further checkout:
                http://pm2.io/


                        -------------

[PM2] Spawning PM2 daemon with pm2_home=/root/.pm2
[PM2] PM2 Successfully daemonized
--> Deploying to dev environment
--> on host 1xx.5xx.x.xx
  ○ hook pre-setup
  ○ running setup
  ○ cloning git@gitlab.com:user/group/repository.git
  ○ full fetch
fatal: destination path 'home/user/source' already exists and is not an empty directory.

  failed to clone

Deploy failed
```

I even tried `- rm -r /home/${TARGET_SERVER_USER_TEST}/source` before setup but it says `rm: can't remove '/home/user/source': No such file or directory`

**Comment by bharadhwaj** [comment]
@noushad-pp I am also facing the same issue. 

How I solved is by,

1. Inside `package.json` , I separated the running command into two parts, **setup** and then **deploy**.

```pm2 deploy ecosystem.config.js <env> setup && pm2 deploy ecosystem.config.js <env> --force```

2. Inside `ecosystem.config.js`, I added a few more hooks like, **pre-setup**, **post-setup** and **post-deploy**

* ```"pre-setup": "rm -rf /home/user/" ```
* ```"post-setup": "<script-to-scp-or-copy-your-env-secret>```
* ``` "post-deploy": "npm install && npm start"```

After this when you run, 
1. Before cloning, it triggers `pre-setup` hook and it will remove the current folder and clone it.
2. After that, it trigger `post-setup` hook and it will move the secrets inside the repo.
3. Now, in the post deploy install node_modules and start the app.


This is a not a proper solution. But this works!
